File: Backend/Group/Type/request.go
package Type

type Request interface {
	GetUserID() string
}

type CreateGroupRequest struct {
	UserID string      `json:"userID" validate:"required"`
	Group  CreateGroup `json:"group" validate:"required"`
}

func (c CreateGroupRequest) GetUserID() string {
	return c.UserID
}

type DeleteGroupRequest struct {
	UserID  string `json:"userID" validate:"required"`
	GroupID string `json:"groupID" validate:"required"`
}

func (d DeleteGroupRequest) GetUserID() string {
	return d.UserID
}

type AddMembersRequest struct {
	UserID  string      `json:"userID" validate:"required"`
	GroupID string      `json:"groupID" validate:"required"`
	Members []AddMember `json:"members" validate:"required"`
}

func (a AddMembersRequest) GetUserID() string {
	return a.UserID
}

type DeleteMemberRequest struct {
	UserID   string `json:"userID"`
	GroupID  string `json:"groupID"`
	MemberID string `json:"memberID"`
}

func (d DeleteMemberRequest) GetUserID() string {
	return d.UserID
}

type JoinGroupRequest struct {
	ID       string `json:"id" validate:"required"` // id of the document in OpenGroups
	UserID   string `json:"userID" validate:"required"`
	GroupID  string `json:"groupID" validate:"required"`
	MemberID string `json:"memberID" validate:"required"`
}

func (j JoinGroupRequest) GetUserID() string {
	return j.UserID
}

type SettleGroupRequest struct {
	UserID  string `json:"userID"`
	GroupID string `json:"groupID"`
}

func (s SettleGroupRequest) GetUserID() string {
	return s.UserID
}

File: Frontend/src/type/request.ts
import type { CreateExpense, TMember, TPayer } from "./type";
import type { currency } from "../const/const";

export interface RegisterRequest {
  name: string;
  account: string;
  password: string;
  confirmPassword: string;
}

export interface LoginRequest {
  account: string;
  password: string;
}

export interface LogoutRequest {
  userID: string;
}

export interface DeleteMemberRequest {
  userID: string;
  groupID: string;
  memberID: string;
}

export interface DeleteExpenseRequest {
  userID: string;
  groupID: string;
  expenseID: string;
}

export interface CreateGroup {
  name: string;
  creatorName: string;
  members: TMember[];
}
export interface CreateGroupRequest {
  userID: string;
  group: CreateGroup;
}

export interface AddMember {
  name: string;
}
export interface AddMemberRequest {
  userID: string;
  groupID: string;
  members: AddMember[];
}

export interface DeleteGroupRequest {
  userID: string;
  groupID: string;
}

export interface CreateExpenseRequest {
  userID: string;
  groupID: string;
  expense: CreateExpense;
}

export interface ExpenseUpdateRequest {
  userID: string;
  groupID: string;
  expenseID: string;
  expenseName?: string;
  amount?: number;
  actualPayerID?: string;
  currency?: (typeof currency)[number];
  split: boolean;
  note?: string;
  remove: string[];
  add: TPayer[];
  update: Partial<TPayer>[];
}

export interface GetJoinGroupLink {
  userID: string;
  groupID: string;
}

export interface JoinGroupRequest {
  id: string; // id of the document in OpenGroups
  userID: string;
  groupID: string;
  memberID: string;
}

export interface ChangeProfileImageRequest {
  userID: string;
}

export interface SettleGroupRequest {
  userID: string;
  groupID: string;
}
